### ISSUE TITLE
proxy: Test that inbound listener refuses connections until its identity is certified

### ISSUE THREAD

**Issue by hawkw** [linkerd/linkerd2#2598]
It would be nice to have a test asserting that a proxy with identity enabled will refuse inbound connections when its identity has yet to be certified. 

We can probably accomplish this pretty easily with the identity test infrastructure added in linkerd/linkerd2-proxy#227

**Comment by olix0r** [comment]
To be clear: a proxy that has not yet received identity will not refuse ALL inbound connections, just those that use TLS and include an SNI value that matches the pods identity.

**Comment by zaharidichev** [comment]
Thanks @olix0r. That makes sense. With the risk of being a bit annoyng, I have to confess that I might need a bit of mentoring on this one. So currently there is a test that asserts that the proxy is not "ready" until its identity is not certified right [here](https://github.com/linkerd/linkerd2-proxy/blob/master/tests/identity.rs#L51). What we need to do is to ensure that inbound TLS connections with an SNI mathing the pod's identity are refused during that window.  However there are two things that are a bit unclear to me.

**Comment by zaharidichev** [comment]
Thanks @olix0r. That makes sense. With the risk of being a bit annoying, I have to confess that I might need a bit of mentoring on this one. So currently there is a test that asserts that the proxy is not "ready" until its identity is not certified right [here](https://github.com/linkerd/linkerd2-proxy/blob/master/tests/identity.rs#L51). What we need to do is to ensure that inbound TLS connections with an SNI ma thing the pod's identity are refused during that window.  However there are two things that are a bit unclear to me: 

- Currently we have a client::http1/http2/tcp that we can use to initiate inbound connections. This client however does not seem to be SSL enabled. Does the scope of the task include adding TLS functionality

- Assuming we have the above in place and we can fire inbound TLS requests, at what point in the code is the "refusal" of this incoming request happening. 

Thanks a bunch in advance.

**Comment by olix0r** [comment]
> Currently we have a client::http1/http2/tcp that we can use to initiate inbound connections. This client however does not seem to be SSL enabled. Does the scope of the task include adding TLS functionality

Yes, probably. That may or may not make sense as its own change (I'm not sure).

> Assuming we have the above in place and we can fire inbound TLS requests, at what point in the code is the "refusal" of this incoming request happening.

I'm not 100% sure that this actually happens today, hence the need for a test ;)

But I'd expect this future to fail, such that the entire connection is just dropped:

https://github.com/linkerd/linkerd2-proxy/blob/61db2e77a247f7b0235b67581f60e8a92f8543cb/src/transport/tls/listen.rs#L353-L357

> Thanks a bunch in advance.

Thanks for getting involved and asking good questions ;)

**Comment by kleimkuhler** [comment]
@zaharidichev Thanks for showing interest in this issue! I just came across the need for this in writing the tests for an issue I am working on: #2676.

I had not updated the issue description much, but I just added some more details that should hopefully add some clarity to your question:

> Assuming we have the above in place and we can fire inbound TLS requests, at what point in the code is the "refusal" of this incoming request happening.

I notice you are on the Linkerd slack, so maybe we can can sync up on there and I can share some thoughts on this issue!
